src/libraries/System.Private.CoreLib/src/System/Threading/Volatile.cs

```diff
@@ -13,176 +13,202 @@ public static class Volatile
         [NonVersionable]
-        public static double Read(ref readonly double location)
-        {
-            long result = Read(ref Unsafe.As<double, long>(ref Unsafe.AsRef(in location)));
-            return BitConverter.Int64BitsToDouble(result);
-        }
+        public static double Read(ref readonly double location) =>
+            BitConverter.Int64BitsToDouble(Read(ref Unsafe.BitCast<double, long>(location)));
 
         [Intrinsic]
         [NonVersionable]
         public static void Write(ref double location, double value) =>
             Write(ref Unsafe.As<double, long>(ref location), BitConverter.DoubleToInt64Bits(value));
-        #endregion
-
-        #region Int16
-        private struct VolatileInt16 { public volatile short Value; }
 
         [Intrinsic]
         [NonVersionable]
-        public static short Read(ref readonly short location) =>
-            Unsafe.As<short, VolatileInt16>(ref Unsafe.AsRef(in location)).Value;
+        public static short Read(ref readonly short location)
+        {
+            short value = location;
+            ReadBarrier();
+            return value;
+        }
 
         [Intrinsic]
         [NonVersionable]
-        public static void Write(ref short location, short value) =>
-            Unsafe.As<short, VolatileInt16>(ref location).Value = value;
-        #endregion
-
-        #region Int32
-        private struct VolatileInt32 { public volatile int Value; }
+        public static void Write(ref short location, short value)
+        {
+            WriteBarrier();
+            location = value;
+        }
 
         [Intrinsic]
         [NonVersionable]
-        public static int Read(ref readonly int location) =>
-            Unsafe.As<int, VolatileInt32>(ref Unsafe.AsRef(in location)).Value;
+        public static int Read(ref readonly int location)
+        {
+            int value = location;
+            ReadBarrier();
+            return value;
+        }
 
         [Intrinsic]
         [NonVersionable]
-        public static void Write(ref int location, int value) =>
-            Unsafe.As<int, VolatileInt32>(ref location).Value = value;
-        #endregion
+        public static void Write(ref int location, int value)
+        {
+            WriteBarrier();
+            location = value;
+        }
 
-        #region Int64
         [Intrinsic]
         [NonVersionable]
-        public static long Read(ref readonly long location) =>
+        public static long Read(ref readonly long location)
+        {
 #if TARGET_64BIT
-            (long)Unsafe.As<long, VolatileIntPtr>(ref Unsafe.AsRef(in location)).Value;
+            long value = location;
+            ReadBarrier();
+            return value;
 #else
             // On 32-bit machines, we use Interlocked, since an ordinary volatile read would not be atomic.
-            Interlocked.CompareExchange(ref Unsafe.AsRef(in location), 0, 0);
+            return Interlocked.CompareExchange(ref Unsafe.AsRef(in location), 0, 0);
```

Comment:
I know this is existing, but this is a bug/memory safety violation.

Consider the case where `location` refers to actually immutable memory. In such a case, `Volatile.Read` should succeed, but in actuality, `CompareExchange` will end up writing `0` in the case the value is actually `0`, which will cause an AV.

Reply:
Aligned 64-bit reads are atomic on xarch. So, `Read` could be implemented as 
```cs
public unsafe long Read(ref long location)
{
    fixed (long* pLocation = &location)
    {
        return Sse2.LoadScalarVector128(pLocation).ToScalar();
    }
}
```

Implementation for Arm32 requires new intrinsic for `ldrexd`, though...

Reply:
> Aligned 64-bit reads

`long ` is not guaranteed to be aligned on 32-bit x86.